### PR TITLE
pybind/mgr/dashboard: fix reverse proxy support

### PR DESCRIPTION
**Pull request by nrdmn**
Changes in 4f7007d break support for reverse http proxies by always redirecting to / on standby MGRs. This patch should fix it.

Fixes: http://tracker.ceph.com/issues/22557
Signed-off-by: Nick Erdmann <n@nirf.de>

### CODE REVIEW

File: src/pybind/mgr/dashboard/module.py [hunk 1]
```diff
@@ -65,12 +65,24 @@ def recurse_refs(root, path):
 def get_prefixed_url(url):
     return global_instance().url_prefix + url
 
+def prepare_url_prefix(url_prefix):
+    if url_prefix == None:
+            url_prefix = ''
+    else:
+        if len(url_prefix) != 0:
+            if url_prefix[0] != '/':
+                url_prefix = '/'+url_prefix
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: could use
```py
url_prefix = urlparse.urljoin('/', url_prefix)
```

File: src/pybind/mgr/dashboard/module.py [hunk 2]
```diff
@@ -65,12 +65,24 @@ def recurse_refs(root, path):
...
+            if url_prefix[-1] == '/':
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: could just put:
```py
url_prefix = url_prefix.rstrip('/')
```

File: src/pybind/mgr/dashboard/module.py [hunk 3]
```diff
@@ -65,12 +65,24 @@ def recurse_refs(root, path):
 def get_prefixed_url(url):
     return global_instance().url_prefix + url
 
+def prepare_url_prefix(url_prefix):
+    if url_prefix == None:
+            url_prefix = ''
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: probably we can use #19948 to simplify this a little bit?

Comment (nrdmn) [reply src/pybind/mgr/dashboard/module.py]: I'll change this as soon as #19948 is merged.

**Comment by nrdmn** [comment]
Thank you for your feedback! I added most of your suggestions, see my other comment.

**Comment by nrdmn** [comment]
@tchaikov I've modified my code to use your improvement from #19948.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-01-25_14:55:14-rados-wip-kefu-testing-2018-01-24-1041-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-01-24_07:50:54-rados-wip-kefu-testing-2018-01-24-1041-distro-basic-mira/

**Comment by tchaikov** [comment]
@jcsp @nrdmn is http://pulpito.ceph.com/kchai-2018-01-25_14:55:14-rados-wip-kefu-testing-2018-01-24-1041-distro-basic-mira/2109202/ relevant ?

**Comment by nrdmn** [comment]
Looks like the test fails because [this method in the tests](https://github.com/ceph/ceph/blob/master/qa/tasks/mgr/mgr_test_case.py#L130) only returns the dashboard's hostname and port, but not the path.

**Comment by jcsp** [comment]
```
2018-01-29 05:07:16,974.974 INFO:__main__:======================================================================
2018-01-29 05:07:16,974.974 INFO:__main__:FAIL: test_standby (tasks.mgr.test_dashboard.TestDashboard)
2018-01-29 05:07:16,974.974 INFO:__main__:----------------------------------------------------------------------
2018-01-29 05:07:16,974.974 INFO:__main__:Traceback (most recent call last):
2018-01-29 05:07:16,974.974 INFO:__main__:  File "/home/jspray/ceph.bravo/qa/tasks/mgr/test_dashboard.py", line 35, in test_standby
2018-01-29 05:07:16,975.975 INFO:__main__:    self.assertEqual(r.headers['Location'], failed_over_uri)
2018-01-29 05:07:16,975.975 INFO:__main__:AssertionError: 'http://senta04.front.sepia.ceph.com:7791/' != u'http://senta04.front.sepia.ceph.com:7791'
```

@tchaikov @nrdmn yeah, that looks like it was broken by this PR -- the call to set_uri on an active mgr should be setting something that matches what the standby redirects to (i.e. both should to have the trailing slash, or neither should)

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-01-28_09:53:35-rados-wip-kefu-testing-2018-01-27-2356-distro-basic-mira/

the failed tests are tracked by

- http://tracker.ceph.com/issues/20086
- http://tracker.ceph.com/issues/21312

**Comment by tchaikov** [comment]
@jcsp as @nrdmn mentioned last week, he's fixed the issue found in last rados qa batch. and it passed this batch. so i guess it's ready to merge.

**Comment by jcsp** [comment]
:+1: :+1: :+1:

**Pull request merged**